test/helpers/kubectl.go

```diff
@@ -1374,3 +1384,188 @@ func (kub *Kubectl) CiliumHealthPreFlightCheck() error {
+					k8sSvc.service.Namespace, k8sSvc.service.Name, pod)
+			}
+			// set found to false for next checks
+			k8sSvc.found = false
+		}
+
+		ciliumLbRes := kub.CiliumExec(pod, ciliumBpfLbCmd)
+		err = ciliumLbRes.GetErr(
+			fmt.Sprintf("Unable to retrieve Cilium bpf lb list on %s", pod))
+		if err != nil {
+			return err
+		}
+
+		var bpfLbMap map[string][]string
+		err = ciliumLbRes.Unmarshal(&bpfLbMap)
+		if err != nil {
+			return fmt.Errorf("Unable to unmarshal Cilium bpf lb list: %s", err.Error())
+		}
+
+		for _, cSvc := range ciliumSvcs {
+			err := validateCiliumSvcLB(cSvc, bpfLbMap, pod)
+			if err != nil {
+				return err
+			}
+		}
+		if len(ciliumSvcs) != len(bpfLbMap) {
+			return fmt.Errorf("Length of Cilium services doesn't match length of bpf LB map on pod %s", pod)
+		}
+	}
+	return nil
+}
+
+// validateCiliumSvc checks if given Cilium service has corresponding k8s services and endpoints in given slices
+// SIDE EFFECT: It marks k8sServices as found if they are matching
+func validateCiliumSvc(cSvc models.Service, k8sSvcs []*k8sServiceFound, k8sEps []v1.Endpoints, pod string) error {
+	var k8sService *k8sServiceFound
+	for _, k8sSvc := range k8sSvcs {
+		if k8sSvc.service.Spec.ClusterIP == cSvc.Status.Realized.FrontendAddress.IP {
+			k8sService = k8sSvc
+			break
+		}
+	}
+	if k8sService == nil {
+		return fmt.Errorf("Could not find Cilium service with ip %s from pod %s in k8s", cSvc.Spec.FrontendAddress.IP, pod)
+	}
+
+	var k8sServicePort *v1.ServicePort
+	for _, k8sPort := range k8sService.service.Spec.Ports {
+		if k8sPort.Port == int32(cSvc.Status.Realized.FrontendAddress.Port) {
+			k8sServicePort = &k8sPort
+			k8sService.found = true
+			break
+		}
+	}
+	if k8sServicePort == nil {
+		return fmt.Errorf("Could not find Cilium service with address %s:%d from pod %s in k8s", cSvc.Spec.FrontendAddress.IP, cSvc.Spec.FrontendAddress.Port, pod)
+	}
+
+	for _, backAddr := range cSvc.Status.Realized.BackendAddresses {
+		foundEp := false
+		for _, k8sEp := range k8sEps {
+			for _, epAddr := range getK8sEndpointAddresses(k8sEp) {
+				if addrsEqual(backAddr, epAddr) {
+					foundEp = true
+				}
+			}
+		}
+		if !foundEp {
+			return fmt.Errorf(
+				"Could not match cilium service backend address %s:%d with k8s endpoint",
+				*backAddr.IP, backAddr.Port)
+		}
+	}
+	return nil
+}
+
+func validateCiliumSvcLB(cSvc models.Service, lbMap map[string][]string, pod string) error {
+	frontendAddress := cSvc.Status.Realized.FrontendAddress.IP + ":" + strconv.Itoa(int(cSvc.Status.Realized.FrontendAddress.Port))
+	bpfBackends, ok := lbMap[frontendAddress]
+	if !ok {
+		return fmt.Errorf("%s bpf lb map entry not found on pod %s", frontendAddress, pod)
+	}
+
+BACKENDS:
+	for _, addr := range cSvc.Status.Realized.BackendAddresses {
+		backend := *addr.IP + ":" + strconv.Itoa(int(addr.Port))
+		for _, bpfAddr := range bpfBackends {
+			if strings.Contains(bpfAddr, backend) {
+				continue BACKENDS
```

Comment:
The thing is I need to iterate over all `bpfBackends` here, dropping the label means adding `found` bool variable here, I think label is clearer and it is aligned with what is in official docs: https://golang.org/ref/spec#Continue_statements

Reply:
Good point.
It occurred to me that if we sorted both lists, the comparision would be simpler. I did something similar during the DNS thing (https://github.com/cilium/cilium/blob/master/pkg/fqdn/dnspoller.go#L398-L419) and it might simplify things? it was a bit more code, vertically, but less nested things (and nominally faster O(nlogn + nlogn + n) vs O(n^2) but these lists are too small for that to matter).